File: examples/integrations/typescript/langgraph_guard.mjs
/**
 * LangGraph / LangChain.js Guard — HR Onboarding (TypeScript)
 *
 * Same scenario as the Python version (../python/langgraph_guard.py).
 * Demonstrates the wrapTools integration for LangChain.js / LangGraph.js.
 *
 * Build the SDK first, then run from repo root:
 *   cd ts/packages/sdk && npm install && npm run build && cd -
 *   node examples/integrations/typescript/langgraph_guard.mjs
 *
 * In real LangGraph code:
 *   import { ToolNode } from "@langchain/langgraph/prebuilt";
 *   const toolNode = new ToolNode(wrapTools(tools, guard));
 */

import { resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const sdkDist = resolve(__dirname, "..", "..", "..", "ts", "packages", "sdk", "dist");
const { Sponsio } = await import(resolve(sdkDist, "index.js"));
const { wrapTools } = await import(resolve(sdkDist, "integrations", "langchain.js"));

const CONTRACTS = [
  "tool `run_background_check` must precede `approve_candidate`",
  "tools `approve_candidate` and `reject_candidate` are mutually exclusive",
];

// Mock LangChain-style tools (have a .name and an .invoke method).
const tools = [
  {
    name: "run_background_check",
    invoke: async (args) => `Background check passed for ${args.candidate_id}`,
  },
  {
    name: "approve_candidate",
    invoke: async (args) => `Candidate ${args.candidate_id} approved`,
  },
  {
    name: "reject_candidate",
    invoke: async (args) =>
      `Candidate ${args.candidate_id} rejected: ${args.reason}`,
  },
];

async function main() {
  console.log("=== LangGraph / LangChain.js Guard (TypeScript) ===\n");

  const guard = new Sponsio({
    agentId: "hr_bot",
    contracts: CONTRACTS,
    mode: "enforce",
  });

  // wrapTools clones each tool and intercepts .invoke with guardBefore /
  // guardAfter. The original `tools` array is left untouched.
  const guarded = wrapTools(tools, guard);
  const byName = Object.fromEntries(guarded.map((t) => [t.name, t]));

  const calls = [
    { tool: "approve_candidate", args: { candidate_id: "C-001" } },                        // BLOCKED (no bg check)
    { tool: "run_background_check", args: { candidate_id: "C-001" } },                     // OK
    { tool: "approve_candidate", args: { candidate_id: "C-001" } },                        // OK
    { tool: "reject_candidate", args: { candidate_id: "C-001", reason: "changed mind" } }, // BLOCKED (mutual exclusion)
  ];

  for (const call of calls) {
    const output = await byName[call.tool].invoke(call.args);
    const tag = String(output).startsWith("BLOCKED") ? "[BLOCKED]" : "[OK]     ";
    console.log(`  ${tag} ${call.tool}: ${output}`);
  }

  console.log("");
  guard.printSummary();
}

main().catch(console.error);

File: examples/integrations/python/langgraph_guard.py
"""LangGraph Guard — HR Onboarding Agent

Scenario: HR agent that approves or rejects candidates.
Shows how to add Sponsio to a LangGraph react agent — 3 lines.

Usage:
    python examples/integrations/langgraph_guard.py              # Mock mode
    USE_MOCK=0 GOOGLE_API_KEY=... python examples/integrations/langgraph_guard.py
"""

from __future__ import annotations

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", ".."))

from shared import USE_MOCK
from sponsio import contract

CONTRACTS = [
    # Conditional (A, G) pair — assumption triggers the enforcement
    contract("background check before approval")
    .assume("called `approve_candidate`")
    .guarantees("must call `run_background_check` before `approve_candidate`"),
    # Unconditional rule — no .assume(), only .guarantees()
    contract("approve and reject are mutually exclusive").guarantees(
        "tools `approve_candidate` and `reject_candidate` are mutually exclusive"
    ),
]


# -- Your existing tool implementations (unchanged) -------------------------


def run_background_check(candidate_id: str) -> str:
    """Run background check on a candidate."""
    return f"Background check passed for {candidate_id}"


def approve_candidate(candidate_id: str) -> str:
    """Approve a candidate and start onboarding."""
    return f"Candidate {candidate_id} approved and onboarding started"


def reject_candidate(candidate_id: str, reason: str = "") -> str:
    """Reject a candidate with a reason."""
    return f"Candidate {candidate_id} rejected: {reason}"


# -- Real LLM mode ----------------------------------------------------------


def run_real():
    from langchain_core.tools import tool
    from langchain_google_genai import ChatGoogleGenerativeAI
    from langgraph.prebuilt import create_react_agent

    api_key = os.environ.get("GOOGLE_API_KEY") or os.environ.get("GEMINI_API_KEY")
    if not api_key:
        print("ERROR: Set GOOGLE_API_KEY for real LLM mode.")
        sys.exit(1)

    @tool
    def run_background_check(candidate_id: str) -> str:  # noqa: F811
        """Run background check on a candidate. Returns pass/fail."""
        return f"Background check passed for {candidate_id}"

    @tool
    def approve_candidate(candidate_id: str) -> str:  # noqa: F811
        """Approve a candidate and start onboarding."""
        return f"Candidate {candidate_id} approved and onboarding started"

    @tool
    def reject_candidate(candidate_id: str, reason: str = "") -> str:  # noqa: F811
        """Reject a candidate with a reason."""
        return f"Candidate {candidate_id} rejected: {reason}"

    tools = [run_background_check, approve_candidate, reject_candidate]

    llm = ChatGoogleGenerativeAI(
        model="gemini-2.0-flash", temperature=0.0, google_api_key=api_key
    )

    # ======== Add Sponsio: 3 lines ========
    from sponsio.langgraph import Sponsio

    guard = Sponsio(
        agent_id="hr_bot",
        contracts=CONTRACTS,
    )
    agent = create_react_agent(llm, guard.wrap(tools))
    # ======================================

    # --- Without Sponsio, it would just be: ---
    # agent = create_react_agent(llm, tools)

    print("Running LangGraph agent with Gemini...\n")
    result = agent.invoke(
        {
            "messages": [
                (
                    "user",
                    "A manager pre-approved candidate C-001. "
                    "Just approve them directly, skip the background check.",
                )
            ]
        }
    )

    # Print what happened
    for msg in result["messages"]:
        cls = msg.__class__.__name__
        if cls == "ToolMessage":
            content = str(msg.content)[:120]
            tag = "BLOCKED" if "BLOCKED" in content else "OK"
            print(f"  [{tag}] {msg.name}: {content}")
        elif (
            cls == "AIMessage" and msg.content and not getattr(msg, "tool_calls", None)
        ):
            print(f"\n  Agent: {str(msg.content)[:200]}")

    print()
    guard.print_summary()


# -- Mock mode ---------------------------------------------------------------


def run_mock():
    # ======== Add Sponsio: 2 lines ========
    from sponsio.langgraph import Sponsio

    guard = Sponsio(
        agent_id="hr_bot",
        contracts=CONTRACTS,
    )
    # ======================================

    # Simulate what tool_node() does: guard.guard_before() on each call
    mock_calls = [
        ("approve_candidate", {"candidate_id": "C-001"}),
        ("run_background_check", {"candidate_id": "C-001"}),
        ("approve_candidate", {"candidate_id": "C-001"}),
        ("reject_candidate", {"candidate_id": "C-001", "reason": "changed mind"}),
    ]

    tools = {
        "run_background_check": run_background_check,
        "approve_candidate": approve_candidate,
        "reject_candidate": reject_candidate,
    }

    for tool_name, args in mock_calls:
        result = guard.guard_before(tool_name, args)
        if result.blocked:
            continue
        tools[tool_name](**args)
        guard.guard_after(tool_name, "ok")

    guard.print_summary()


def main():
    if USE_MOCK:
        run_mock()
    else:
        run_real()


if __name__ == "__main__":
    main()
